File: library/src/main/java/com/yy/codex/uikit/UIFont.java
package com.yy.codex.uikit;

import android.graphics.Typeface;
import android.support.annotation.Nullable;

import static android.graphics.Typeface.DEFAULT;
import static android.graphics.Typeface.DEFAULT_BOLD;

/**
 * Created by cuiminghui on 2017/1/9.
 */

public class UIFont {

    @Nullable
    final public String fontFamily;
    final public float  fontSize;

    static public UIFont systemBold(float fontSize) {
        return new UIFont("SystemBold", fontSize);
    }

    public UIFont(float fontSize) {
        this.fontFamily = "System";
        this.fontSize = fontSize;
    }

    public UIFont(@Nullable String fontFamily, float fontSize) {
        this.fontFamily = fontFamily;
        this.fontSize = fontSize;
    }

    @Nullable
    public Typeface getTypeface() {
        if (fontFamily.equalsIgnoreCase("System")) {
            return DEFAULT;
        }
        else if (fontFamily.equalsIgnoreCase("SystemBold")) {
            return DEFAULT_BOLD;
        }
        return Typeface.create(fontFamily, Typeface.NORMAL);
    }

}

File: library/src/main/java/com/yy/codex/uikit/UIFont.kt
package com.yy.codex.uikit

import android.graphics.Typeface

import android.graphics.Typeface.DEFAULT
import android.graphics.Typeface.DEFAULT_BOLD
import android.text.TextPaint

/**
 * Created by cuiminghui on 2017/1/9.
 */

class UIFont(val fontFamily: String?, val fontSize: Float) {

    constructor(fontSize: Float): this(null, fontSize)

    val typeface: Typeface?
        get() {
            val fontFamily = fontFamily ?: return Typeface.create(fontFamily, Typeface.NORMAL)
            if (fontFamily.equals("System", ignoreCase = true)) {
                return DEFAULT
            }
            else if (fontFamily.equals("SystemBold", ignoreCase = true)) {
                return DEFAULT_BOLD
            }
            else {
                return Typeface.create(fontFamily, Typeface.NORMAL)
            }
        }

    private val textPaint = TextPaint()

    var lineHeight: Double = 0.0
        get() {
            typeface?.let {
                textPaint.typeface = it
                textPaint.textSize = (fontSize * UIScreen.mainScreen.scale()).toFloat()
                return (Math.abs(textPaint.fontMetrics.ascent).toDouble() + Math.abs(textPaint.fontMetrics.descent).toDouble()) / UIScreen.mainScreen.scale()
            }
            return 0.0
        }

    companion object {

        fun systemBold(fontSize: Float): UIFont {
            return UIFont("SystemBold", fontSize)
        }

    }

}
